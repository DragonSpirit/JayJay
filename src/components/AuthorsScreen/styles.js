import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  authorInput: {
    fontSize: 20,
    textAlign: 'center',
    margin: 20,
    height: 40,
    width: '80%',
  },
  authorsContainer: {
    flex: 1,
    paddingTop: 30,
  },
})

export default styles
