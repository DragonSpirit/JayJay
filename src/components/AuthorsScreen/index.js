// @flow

import React from 'react'
import {
  View,
  TextInput,
  FlatList,
  Alert,
} from 'react-native'
import { AuthorsTabIcon } from '../TabIcons/AuthorsTabIcon'
import { array, func, bool } from 'prop-types'
import { Button, ListItem, Icon } from 'react-native-elements'
import styles from './styles'
import { commonStyles } from '../common.styles'

type Props = {
  authors: Array<string>,
  fetchAuthors: (author: string) => any,
  deleteAuthor: (author: string) => any,
  fetchPosts: (author: string) => any,
  authorsLoading: boolean,
}

type State = {
  name: string,
}

class AuthorsScreen extends React.PureComponent<Props, State> {

  static navigationOptions = {
    title: 'Авторы',
    tabBarIcon: AuthorsTabIcon,
  }

  static propTypes = {
    authors: array,
    fetchAuthors: func,
    deleteAuthor: func,
    fetchPosts: func,
    authorsLoading: bool,
  };

  static defaultProps = {
    authors: [],
    fetchAuthors: () => {},
    deleteAuthor: () => {},
    authorsLoading: false,
  }

  constructor(props: Props) {
    super(props)
  }

  state: State = {
    name: '',
  }

  addUser = (): void => {
    this.props.fetchAuthors(this.state.name)
      .then(this.props.fetchPosts(this.state.name))
      .catch((error: Error) => {
        switch (error.message) {
          case 'unknown user':
            Alert.alert(
              'Ошибка',
              'Такого пользователя скорее всего не существует',
            )
            break
          case 'duplicate users':
            Alert.alert(
              'Ошибка',
              'Такой пользователь уже находится в списке отслеживаемых',
            )
            break
          default:
            Alert.alert(
              'Ошибка',
              'Неизвестная ошибка, попробуйте позже',
            )
            break
        }
      })
  }

  _isEmpty = (): boolean => this.state.name.trim().length === 0

  deleteUser = (author: string) => {
    this.props.deleteAuthor(author)
  }

  emptyAuthorView = () => (
    <View style={commonStyles.container}>
      <TextInput
        placeholder='Username'
        onChangeText={name => this.setState({name})}
        style={styles.authorInput}
      />
      <Button title='Добавить пользователя'
        onPress={this.addUser}
        buttonStyle={[commonStyles.buttonStyleCommon, commonStyles.buttonStyleBig]}
        loading={this.props.authorsLoading}
        disabled={this._isEmpty()} />
    </View>
  )

  renderAuthorsItem = (data: Object) => (
    <ListItem
      title={data.item}
      rightIcon={
        <Icon name='delete'
          color='#000'
          onPress={() => this.deleteUser(data.item)}
        />
      }
    />
  )

  _keyExtractor = (item, index): string => `${index}`

  authorsListView = (): any => (
    <View style={styles.authorsContainer}>
      <View style={commonStyles.flexRow}>
        <TextInput
          placeholder='Username'
          onChangeText={(name: string) => this.setState({name})}
          style={[styles.welcome, commonStyles.flex]}
        />
        <Button title='+'
          onPress={this.addUser}
          buttonStyle={[commonStyles.buttonStyleCommon]}
          loading={this.props.authorsLoading}
          disabled={this._isEmpty()} />
      </View>
      <FlatList
        keyExtractor={this._keyExtractor}
        data={this.props.authors}
        renderItem={this.renderAuthorsItem}
        keyboardShouldPersistTaps={'always'}
      />
    </View>
  )

  render() {
    const {
      authors,
    } = this.props
    if (authors.length === 0) {
      return this.emptyAuthorView()
    }
    return this.authorsListView()
  }
}

export default AuthorsScreen
