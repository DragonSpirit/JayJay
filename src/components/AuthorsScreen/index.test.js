import React from 'react';
import AuthorsScreen from './';

import renderer from 'react-test-renderer';

test('empty renders correctly', () => {
  const tree = renderer.create(<AuthorsScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('non empty renders correctly', () => {
  const tree = renderer.create(<AuthorsScreen authors={['test']} />).toJSON();
  expect(tree).toMatchSnapshot();
})
