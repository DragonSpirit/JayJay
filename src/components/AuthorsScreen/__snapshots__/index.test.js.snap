// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Журнал"
    style={
      Object {
        "fontSize": 20,
        "height": 40,
        "margin": 20,
        "textAlign": "center",
        "width": "80%",
      }
    }
  />
  <View
    style={
      Object {
        "borderRadius": 3,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "hsl(208, 8%, 90%)",
            "borderColor": "transparent",
            "borderRadius": 5,
            "borderWidth": 0,
            "flexDirection": "row",
            "height": 45,
            "justifyContent": "center",
            "padding": 8,
            "width": 300,
          }
        }
      >
        <Text
          style={
            Object {
              "color": Object {
                "color": Array [
                  208,
                  8,
                  63,
                ],
                "model": "hsl",
                "valpha": 1,
              },
              "fontSize": 18,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Добавить пользователя
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`non empty renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 30,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Добавить журнал"
      style={
        Array [
          undefined,
          Object {
            "flex": 1,
          },
        ]
      }
    />
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "hsl(208, 8%, 90%)",
              "borderColor": "transparent",
              "borderRadius": 5,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": Object {
                  "color": Array [
                    208,
                    8,
                    63,
                  ],
                  "model": "hsl",
                  "valpha": 1,
                },
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            +
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        "test",
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="always"
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "fontSize": 17,
                  }
                }
                testID="listItemTitle"
                theme={
                  Object {
                    "colors": Object {
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": Object {
                        "android": Object {
                          "error": "#f44336",
                          "primary": "#2196f3",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "ios": Object {
                          "error": "#ff3b30",
                          "primary": "#007aff",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#8F0CE8",
                      "success": "#52c41a",
                      "warning": "#faad14",
                    },
                  }
                }
              >
                test
              </Text>
            </View>
            <View
              width={16}
            />
            <View>
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#000",
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
